experiments/storm/src/main/scala/io/gearpump/experiments/storm/util/StormUtil.scala

```diff
@@ -75,4 +72,205 @@ object StormUtil {
+        val inputStreamId = globalStreamId.get_streamId()
+        val inputComponentId = globalStreamId.get_componentId
+        if (inputComponentId.equals(componentId)) {
+          val curr = allTargets.getOrElse(otherComponentId, Map.empty[String, Grouping])
+          allTargets + (otherComponentId -> (curr + (inputStreamId -> grouping)))
+        } else {
+          allTargets
+        }
+    }
+  }
+
+  def getComponentToSortedTasks(taskToComponent: Map[Integer, String]): JMap[String, JList[Integer]] = {
+    taskToComponent.groupBy(_._2).map { case (component, map) =>
+      component -> map.toList.map(_._1).sorted.asJava }.asJava
+  }
+
+  def getInputs(componentId: String, topology: StormTopology): JMap[GlobalStreamId, Grouping] = {
+    ThriftTopologyUtils.getComponentCommon(topology, componentId).get_inputs
+  }
+
+  def getGrouper(topologyContext: TopologyContext, grouping: Grouping,
+      source: String, streamId: String, target: String): Grouper = {
+    val outFields = topologyContext.getComponentOutputFields(source, streamId)
+    val sourceTasks = topologyContext.getComponentTasks(source)
+    val targetTaskNum = topologyContext.getComponentTasks(target).size
+    val globalStreamId = new GlobalStreamId(source, streamId)
+
+    grouping.getSetField match {
+      case Grouping._Fields.FIELDS =>
+        if (isGlobalGrouping(grouping)) {
+          new GlobalGrouper
+        } else {
+          new FieldsGrouper(outFields, new Fields(grouping.get_fields()), targetTaskNum)
+        }
+      case Grouping._Fields.SHUFFLE =>
+        new ShuffleGrouper(targetTaskNum)
+      case Grouping._Fields.NONE =>
+        new NoneGrouper(targetTaskNum)
+      case Grouping._Fields.ALL =>
+        new AllGrouper(targetTaskNum)
+      case Grouping._Fields.CUSTOM_SERIALIZED =>
+        val customGrouping = Utils.deserialize(grouping.get_custom_serialized()).asInstanceOf[CustomStreamGrouping]
+        val grouper = new CustomGrouper(customGrouping)
+        grouper.prepare(topologyContext, globalStreamId, sourceTasks)
+        grouper
+      case Grouping._Fields.CUSTOM_OBJECT =>
+        val customObject = grouping.get_custom_object()
+        val customGrouping = instantiateJavaObject(customObject)
+        val grouper = new CustomGrouper(customGrouping)
+        grouper.prepare(topologyContext, globalStreamId, sourceTasks)
+        grouper
+      case Grouping._Fields.LOCAL_OR_SHUFFLE =>
+        // GearPump has built-in support for sending messages to local actor
+        new ShuffleGrouper(targetTaskNum)
+      case Grouping._Fields.DIRECT =>
+        throw new Exception("direct grouping should not be called here")
+    }
+  }
+
+  def isGlobalGrouping(grouping: Grouping): Boolean = {
+    grouping.getSetField == Grouping._Fields.FIELDS &&
+        grouping.get_fields.isEmpty
+  }
+
+  def getParallelism(component: ComponentCommon): Int = {
+    val componentConf = parseJsonStringToMap(component.get_json_conf)
```

Comment:
This part seems unclear to me, I don't know what ComponentCommon means, and what the schema of the json_conf used in this function.

I think we will want some clear "line" between storm part and gearpump part. 

One way is to define a wrapper for StormTopology, like GearpumpStormTopology(storm: StormTopology) {
  def raw: StormTolology
}

and define some useful function name that is more clear in this context.